File: vane-trifles/src/main/java/org/oddlama/vane/trifles/items/File.java
package org.oddlama.vane.trifles.items;

import static org.oddlama.vane.util.BlockUtil.raytraceDominantFace;
import static org.oddlama.vane.util.BlockUtil.raytraceOct;
import static org.oddlama.vane.util.ItemUtil.damageItem;
import static org.oddlama.vane.util.PlayerUtil.swingArm;

import java.util.ArrayList;
import java.util.Collections;
import java.util.EnumSet;
import java.util.Set;

import org.bukkit.Material;
import org.bukkit.Sound;
import org.bukkit.SoundCategory;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.data.Bisected;
import org.bukkit.block.data.Directional;
import org.bukkit.block.data.MultipleFacing;
import org.bukkit.block.data.type.Fence;
import org.bukkit.block.data.type.GlassPane;
import org.bukkit.block.data.type.Stairs;
import org.bukkit.block.data.type.Tripwire;
import org.bukkit.block.data.type.Wall;
import org.bukkit.entity.Player;
import org.bukkit.event.Event;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.block.Action;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.oddlama.vane.annotation.item.VaneItem;
import org.oddlama.vane.core.config.recipes.RecipeList;
import org.oddlama.vane.core.config.recipes.ShapedRecipeDefinition;
import org.oddlama.vane.core.item.CustomItem;
import org.oddlama.vane.core.item.api.InhibitBehavior;
import org.oddlama.vane.core.module.Context;
import org.oddlama.vane.trifles.Trifles;
import org.oddlama.vane.util.BlockUtil;

@VaneItem(
    name = "file",
    base = Material.WARPED_FUNGUS_ON_A_STICK,
    durability = 4000,
    modelData = 0x760003,
    version = 1
)
public class File extends CustomItem<Trifles> {

    public File(Context<Trifles> context) {
        super(context);
    }

    @Override
    public RecipeList defaultRecipes() {
        return RecipeList.of(
            new ShapedRecipeDefinition("generic")
                .shape(" M", "S ")
                .setIngredient('M', Material.IRON_INGOT)
                .setIngredient('S', Material.STICK)
                .result(key().toString())
        );
    }

    @Override
    public EnumSet<InhibitBehavior> inhibitedBehaviors() {
        return EnumSet.of(InhibitBehavior.USE_IN_VANILLA_RECIPE, InhibitBehavior.TEMPT, InhibitBehavior.USE_OFFHAND);
    }

    private BlockFace nextFacing(Set<BlockFace> allowedFaces, BlockFace face) {
        if (allowedFaces.isEmpty()) {
            return face;
        }
        final var list = new ArrayList<BlockFace>(allowedFaces);
        Collections.sort(list, (a, b) -> a.ordinal() - b.ordinal());
        final var index = list.indexOf(face);
        if (index == -1) {
            return face;
        }
        return list.get((index + 1) % list.size());
    }

    private Sound changeStairHalf(final Stairs stairs) {
        // Change half
        stairs.setHalf(stairs.getHalf() == Bisected.Half.BOTTOM ? Bisected.Half.TOP : Bisected.Half.BOTTOM);
        return Sound.UI_STONECUTTER_TAKE_RESULT;
    }

    private BlockFace nextFaceCcw(final BlockFace face) {
        switch (face) {
            default:
                return null;
            case NORTH:
                return BlockFace.WEST;
            case EAST:
                return BlockFace.NORTH;
            case SOUTH:
                return BlockFace.EAST;
            case WEST:
                return BlockFace.SOUTH;
        }
    }

    private Sound changeStairShape(
        final Player player,
        final Block block,
        final Stairs stairs,
        final BlockFace clickedFace
    ) {
        // Check which eighth of the block was clicked
        final var oct = raytraceOct(player, block);
        if (oct == null) {
            return null;
        }

        var corner = oct.corner();
        final var isStairTop = stairs.getHalf() == Bisected.Half.TOP;

        // If we clicked on the base part, toggle the corner upstate,
        // as we always want to reason abount the corner to add/remove.
        if (corner.up() == isStairTop) {
            corner = corner.up(!corner.up());
        }

        // Rotate a corner so that we can interpret it from
        // the reference facing (north)
        final var originalFacing = stairs.getFacing();
        corner = corner.rotateToNorthReference(originalFacing);

        // Determine the resulting shape, face and wether the oct got added or removed
        Stairs.Shape shape = null;
        BlockFace face = null;
        boolean added = false;
        switch (stairs.getShape()) {
            case STRAIGHT:
                switch (corner.xzFace()) {
                    case SOUTH_WEST:
                        shape = Stairs.Shape.INNER_LEFT;
                        face = BlockFace.NORTH;
                        added = true;
                        break;
                    case NORTH_WEST:
                        shape = Stairs.Shape.OUTER_RIGHT;
                        face = BlockFace.NORTH;
                        added = false;
                        break;
                    case NORTH_EAST:
                        shape = Stairs.Shape.OUTER_LEFT;
                        face = BlockFace.NORTH;
                        added = false;
                        break;
                    case SOUTH_EAST:
                        shape = Stairs.Shape.INNER_RIGHT;
                        face = BlockFace.NORTH;
                        added = true;
                        break;
                    default:
                        break;
                }
                break;
            case INNER_LEFT:
                switch (corner.xzFace()) {
                    case SOUTH_WEST:
                        shape = Stairs.Shape.STRAIGHT;
                        face = BlockFace.NORTH;
                        added = false;
                        break;
                    case NORTH_EAST:
                        shape = Stairs.Shape.STRAIGHT;
                        face = BlockFace.WEST;
                        added = false;
                        break;
                    default:
                        break;
                }
                break;
            case INNER_RIGHT:
                switch (corner.xzFace()) {
                    case NORTH_WEST:
                        shape = Stairs.Shape.STRAIGHT;
                        face = BlockFace.EAST;
                        added = false;
                        break;
                    case SOUTH_EAST:
                        shape = Stairs.Shape.STRAIGHT;
                        face = BlockFace.NORTH;
                        added = false;
                        break;
                    default:
                        break;
                }
                break;
            case OUTER_LEFT:
                switch (corner.xzFace()) {
                    case SOUTH_WEST:
                        shape = Stairs.Shape.STRAIGHT;
                        face = BlockFace.WEST;
                        added = true;
                        break;
                    case NORTH_EAST:
                        shape = Stairs.Shape.STRAIGHT;
                        face = BlockFace.NORTH;
                        added = true;
                        break;
                    default:
                        break;
                }
                break;
            case OUTER_RIGHT:
                switch (corner.xzFace()) {
                    case NORTH_WEST:
                        shape = Stairs.Shape.STRAIGHT;
                        face = BlockFace.NORTH;
                        added = true;
                        break;
                    case SOUTH_EAST:
                        shape = Stairs.Shape.STRAIGHT;
                        face = BlockFace.EAST;
                        added = true;
                        break;
                    default:
                        break;
                }
                break;
        }

        // Break if the resulting shape is invalid
        if (shape == null) {
            return null;
        }

        // Undo reference rotation
        switch (face) {
            case NORTH:
                face = originalFacing;
                break;
            case EAST:
                face = nextFaceCcw(originalFacing).getOppositeFace();
                break;
            case SOUTH:
                face = originalFacing.getOppositeFace();
                break;
            case WEST:
                face = nextFaceCcw(originalFacing);
                break;
            default:
                break;
        }

        stairs.setShape(shape);
        stairs.setFacing(face);

        return added ? Sound.UI_STONECUTTER_TAKE_RESULT : Sound.BLOCK_GRINDSTONE_USE;
    }

    private Sound changeMultipleFacing(
        final Player player,
        final Block block,
        final MultipleFacing mf,
        BlockFace clickedFace
    ) {
        final int minFaces;
        if (mf instanceof Fence || mf instanceof GlassPane) {
            // Allow fences and glass panes to have 0 faces
            minFaces = 0;

            // Trace which side is the dominant side
            final var result = raytraceDominantFace(player, block);
            if (result == null) {
                return null;
            }

            // Only replace facing choice if we did hit a side,
            // or if the dominance was big enough.
            if (result.dominance > .2 || (clickedFace != BlockFace.UP && clickedFace != BlockFace.DOWN)) {
                clickedFace = result.face;
            }
        } else if (mf instanceof Tripwire) {
            minFaces = 0;
        } else {
            return null;
        }

        // Check if the clicked face is allowed to change
        if (!mf.getAllowedFaces().contains(clickedFace)) {
            return null;
        }

        boolean hasFace = mf.hasFace(clickedFace);
        if (hasFace && minFaces >= mf.getFaces().size()) {
            // Refuse to remove beyond minimum face count
            return null;
        }

        // Toggle clicked block face
        mf.setFace(clickedFace, !hasFace);

        // Choose sound
        return hasFace ? Sound.BLOCK_GRINDSTONE_USE : Sound.UI_STONECUTTER_TAKE_RESULT;
    }

    private Sound changeWall(final Player player, final Block block, final Wall wall, final BlockFace clickedFace) {
        // Trace which side is the dominant side
        final var result = raytraceDominantFace(player, block);
        if (result == null) {
            return null;
        }

        final BlockFace adjustedClickedFace;
        // Only replace facing choice if we did hit a side,
        // or if the dominance was big enough.
        if (result.dominance > .2 || (clickedFace != BlockFace.UP && clickedFace != BlockFace.DOWN)) {
            adjustedClickedFace = result.face;
        } else {
            adjustedClickedFace = clickedFace;
        }

        if (clickedFace == BlockFace.UP) {
            final var wasUp = wall.isUp();
            if (adjustedClickedFace == BlockFace.UP) {
                // click top in middle -> toggle up
                wall.setUp(!wasUp);
            } else if (BlockUtil.XZ_FACES.contains(adjustedClickedFace)) {
                // click top on side -> toggle height
                final var height = wall.getHeight(adjustedClickedFace);
                switch (height) {
                    case NONE:
                        return null;
                    case LOW:
                        wall.setHeight(adjustedClickedFace, Wall.Height.TALL);
                        break;
                    case TALL:
                        wall.setHeight(adjustedClickedFace, Wall.Height.LOW);
                        break;
                }
            }

            return wasUp ? Sound.BLOCK_GRINDSTONE_USE : Sound.UI_STONECUTTER_TAKE_RESULT;
        } else {
            // click side -> toggle side
            final var hasFace = wall.getHeight(adjustedClickedFace) != Wall.Height.NONE;

            // Set height and choose sound
            if (hasFace) {
                wall.setHeight(adjustedClickedFace, Wall.Height.NONE);
                return Sound.BLOCK_GRINDSTONE_USE;
            } else {
                // Use opposite face's height, or low if there is nothing.
                var targetHeight = wall.getHeight(adjustedClickedFace.getOppositeFace());
                if (targetHeight == Wall.Height.NONE) {
                    targetHeight = Wall.Height.LOW;
                }
                wall.setHeight(adjustedClickedFace, targetHeight);
                return Sound.UI_STONECUTTER_TAKE_RESULT;
            }
        }
    }

    private Sound changeDirectionalFacing(
        final Player player,
        final Block block,
        final Directional directional,
        final BlockFace clickedFace
    ) {
        // Toggle facing
        directional.setFacing(nextFacing(directional.getFaces(), directional.getFacing()));
        return Sound.UI_STONECUTTER_TAKE_RESULT;
    }

    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = true)
    public void onPlayerRightClick(final PlayerInteractEvent event) {
        if (!event.hasBlock() || event.getAction() != Action.RIGHT_CLICK_BLOCK) {
            return;
        }

        if (event.useItemInHand() == Event.Result.DENY) {
            return;
        }

        // Get item variant
        final var player = event.getPlayer();
        final var item = player.getEquipment().getItem(event.getHand());
        if (!isInstance(item)) {
            return;
        }

        // Create a block break event for block to transmute and check if it gets canceled
        final var block = event.getClickedBlock();
        final var breakEvent = new BlockBreakEvent(block, player);
        getModule().getServer().getPluginManager().callEvent(breakEvent);
        if (breakEvent.isCancelled()) {
            return;
        }

        final var data = block.getBlockData();
        final var clickedFace = event.getBlockFace();

        final Sound sound;
        if (player.isSneaking()) {
            if (data instanceof Stairs) {
                sound = changeStairHalf((Stairs) data);
            } else {
                return;
            }
        } else {
            if (data instanceof MultipleFacing) {
                sound = changeMultipleFacing(player, block, (MultipleFacing) data, clickedFace);
            } else if (data instanceof Wall) {
                sound = changeWall(player, block, (Wall) data, clickedFace);
            } else if (data instanceof Stairs) {
                sound = changeStairShape(player, block, (Stairs) data, clickedFace);
            } else {
                return;
            }
        }

        // Return if nothing was done
        if (sound == null) {
            return;
        }

        // Update block data, and don't trigger physics! (We do not want to affect surrounding
        // blocks!)
        block.setBlockData(data, false);
        block.getWorld().playSound(block.getLocation(), sound, SoundCategory.BLOCKS, 1.0f, 1.0f);

        // Damage item and swing arm
        damageItem(player, item, 1);
        swingArm(player, event.getHand());
    }
}

File: vane-trifles/src/main/kotlin/org/oddlama/vane/trifles/items/File.kt
package org.oddlama.vane.trifles.items

import org.bukkit.Material
import org.bukkit.Sound
import org.bukkit.SoundCategory
import org.bukkit.block.Block
import org.bukkit.block.BlockFace
import org.bukkit.block.data.Bisected
import org.bukkit.block.data.MultipleFacing
import org.bukkit.block.data.type.*
import org.bukkit.block.data.type.Tripwire
import org.bukkit.entity.Player
import org.bukkit.event.Event
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.block.Action
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.oddlama.vane.annotation.item.VaneItem
import org.oddlama.vane.core.config.recipes.RecipeList
import org.oddlama.vane.core.config.recipes.ShapedRecipeDefinition
import org.oddlama.vane.core.item.CustomItem
import org.oddlama.vane.core.item.api.InhibitBehavior
import org.oddlama.vane.core.module.Context
import org.oddlama.vane.trifles.Trifles
import org.oddlama.vane.util.BlockUtil
import org.oddlama.vane.util.ItemUtil
import org.oddlama.vane.util.PlayerUtil
import java.util.*

@VaneItem(name = "file", base = Material.WARPED_FUNGUS_ON_A_STICK, durability = 4000, modelData = 0x760003, version = 1)
/**
 * Utility tool that cycles block-state variants for stairs, walls, and multi-facing blocks.
 */
class File(context: Context<Trifles?>) : CustomItem<Trifles?>(context) {
    /** Defines the file crafting recipe. */
    override fun defaultRecipes(): RecipeList {
        return RecipeList.of(
            ShapedRecipeDefinition("generic")
                .shape(" M", "S ")
                .setIngredient('M', Material.IRON_INGOT)
                .setIngredient('S', Material.STICK)
                .result(key().toString())
        )
    }

    /** Prevents conflicting vanilla interactions for the file item. */
    override fun inhibitedBehaviors(): EnumSet<InhibitBehavior> {
        return EnumSet.of(
            InhibitBehavior.USE_IN_VANILLA_RECIPE,
            InhibitBehavior.TEMPT,
            InhibitBehavior.USE_OFFHAND
        )
    }

    /** Returns the next allowed facing from a sorted list of valid facings. */
    private fun nextFacing(allowedFaces: MutableSet<BlockFace?>, face: BlockFace?): BlockFace? {
        if (allowedFaces.isEmpty()) {
            return face
        }
        val list = allowedFaces.filterNotNull().sortedBy { it.ordinal }
        if (list.isEmpty()) {
            return face
        }
        val index = list.indexOf(face)
        if (index == -1) {
            return face
        }
        return list[(index + 1) % list.size]
    }

    /** Toggles stair half between top and bottom. */
    private fun changeStairHalf(stairs: Stairs): Sound? {
        // Change half
        stairs.half = if (stairs.half == Bisected.Half.BOTTOM) Bisected.Half.TOP else Bisected.Half.BOTTOM
        return Sound.UI_STONECUTTER_TAKE_RESULT
    }

    /** Rotates a cardinal face counter-clockwise. */
    private fun nextFaceCcw(face: BlockFace): BlockFace {
        return when (face) {
            BlockFace.NORTH -> BlockFace.WEST
            BlockFace.EAST -> BlockFace.NORTH
            BlockFace.SOUTH -> BlockFace.EAST
            BlockFace.WEST -> BlockFace.SOUTH
            else -> face
        }
    }

    /** Cycles stair shape based on clicked octant and current stair orientation. */
    private fun changeStairShape(player: Player, block: Block?, stairs: Stairs): Sound? {
        // Check which eighth of the block was clicked
        val oct = BlockUtil.raytraceOct(player, block) ?: return null

        var corner = oct.corner()
        val isStairTop = stairs.half == Bisected.Half.TOP

        // If we clicked on the base part, toggle the corner upstate,
        // as we always want to reason abount the corner to add/remove.
        if (corner.up() == isStairTop) {
            corner = corner.up(!corner.up())
        }

        // Rotate a corner so that we can interpret it from
        // the reference facing (north)
        val originalFacing = stairs.facing
        corner = corner.rotateToNorthReference(originalFacing)

        // Determine resulting shape/facing and whether geometry is added or removed.
        var shape: Stairs.Shape? = null
        var face: BlockFace? = null
        var added = false
        when (stairs.shape) {
            Stairs.Shape.STRAIGHT -> when (corner.xzFace()) {
                BlockFace.SOUTH_WEST -> {
                    shape = Stairs.Shape.INNER_LEFT
                    face = BlockFace.NORTH
                    added = true
                }

                BlockFace.NORTH_WEST -> {
                    shape = Stairs.Shape.OUTER_RIGHT
                    face = BlockFace.NORTH
                    added = false
                }

                BlockFace.NORTH_EAST -> {
                    shape = Stairs.Shape.OUTER_LEFT
                    face = BlockFace.NORTH
                    added = false
                }

                BlockFace.SOUTH_EAST -> {
                    shape = Stairs.Shape.INNER_RIGHT
                    face = BlockFace.NORTH
                    added = true
                }

                else -> {}
            }

            Stairs.Shape.INNER_LEFT -> when (corner.xzFace()) {
                BlockFace.SOUTH_WEST -> {
                    shape = Stairs.Shape.STRAIGHT
                    face = BlockFace.NORTH
                    added = false
                }

                BlockFace.NORTH_EAST -> {
                    shape = Stairs.Shape.STRAIGHT
                    face = BlockFace.WEST
                    added = false
                }

                else -> {}
            }

            Stairs.Shape.INNER_RIGHT -> when (corner.xzFace()) {
                BlockFace.NORTH_WEST -> {
                    shape = Stairs.Shape.STRAIGHT
                    face = BlockFace.EAST
                    added = false
                }

                BlockFace.SOUTH_EAST -> {
                    shape = Stairs.Shape.STRAIGHT
                    face = BlockFace.NORTH
                    added = false
                }

                else -> {}
            }

            Stairs.Shape.OUTER_LEFT -> when (corner.xzFace()) {
                BlockFace.SOUTH_WEST -> {
                    shape = Stairs.Shape.STRAIGHT
                    face = BlockFace.WEST
                    added = true
                }

                BlockFace.NORTH_EAST -> {
                    shape = Stairs.Shape.STRAIGHT
                    face = BlockFace.NORTH
                    added = true
                }

                else -> {}
            }

            Stairs.Shape.OUTER_RIGHT -> when (corner.xzFace()) {
                BlockFace.NORTH_WEST -> {
                    shape = Stairs.Shape.STRAIGHT
                    face = BlockFace.NORTH
                    added = true
                }

                BlockFace.SOUTH_EAST -> {
                    shape = Stairs.Shape.STRAIGHT
                    face = BlockFace.EAST
                    added = true
                }

                else -> {}
            }
        }

        // Break if the resulting shape is invalid
        if (shape == null) {
            return null
        }

        // Undo reference rotation
        when (face) {
            BlockFace.NORTH -> face = originalFacing
            BlockFace.EAST -> face = nextFaceCcw(originalFacing).oppositeFace
            BlockFace.WEST -> face = nextFaceCcw(originalFacing)
            else -> {}
        }

        stairs.shape = shape
        stairs.facing = face ?: return null

        return if (added) Sound.UI_STONECUTTER_TAKE_RESULT else Sound.BLOCK_GRINDSTONE_USE
    }

    /** Toggles side connections on multiple-facing blocks and returns a feedback sound. */
    private fun changeMultipleFacing(
        player: Player,
        block: Block?,
        mf: MultipleFacing?,
        clickedFace: BlockFace
    ): Sound? {
        var clickedFace = clickedFace
        if (mf is Fence || mf is GlassPane) {
            // Trace which side is the dominant side
            val result = BlockUtil.raytraceDominantFace(player, block) ?: return null

            // Only replace facing choice if we did hit a side,
            // or if the dominance was big enough.
            if (result.dominance > .2 || (clickedFace != BlockFace.UP && clickedFace != BlockFace.DOWN)) {
                clickedFace = result.face ?: return null
            }
        } else if (mf !is Tripwire) {
            return null
        }

        // Check if the clicked face is allowed to change
        if (!mf.allowedFaces.contains(clickedFace)) {
            return null
        }

        val hasFace = mf.hasFace(clickedFace)
        if (hasFace && mf.faces.isEmpty()) {
            // Refuse to remove beyond minimum face count
            return null
        }

        // Toggle clicked block face
        mf.setFace(clickedFace, !hasFace)

        // Choose sound
        return if (hasFace) Sound.BLOCK_GRINDSTONE_USE else Sound.UI_STONECUTTER_TAKE_RESULT
    }

    /** Adjusts wall post/side heights based on click position and dominant hit face. */
    private fun changeWall(player: Player, block: Block?, wall: Wall, clickedFace: BlockFace?): Sound? {
        // Trace which side is the dominant side
        val result = BlockUtil.raytraceDominantFace(player, block) ?: return null
        // Only replace facing choice if we did hit a side,
        // or if the dominance was big enough.
        val adjustedClickedFace =
            if (result.dominance > .2 || (clickedFace != BlockFace.UP && clickedFace != BlockFace.DOWN)) {
                result.face
            } else {
                clickedFace
            }

        if (clickedFace == BlockFace.UP) {
            val wasUp = wall.isUp
            if (adjustedClickedFace == BlockFace.UP) {
                // click top in middle -> toggle up
                wall.isUp = !wasUp
            } else if (BlockUtil.XZ_FACES.contains(adjustedClickedFace)) {
                // click top on side -> toggle height
                val sideFace = adjustedClickedFace ?: return null
                val height = wall.getHeight(sideFace)
                when (height) {
                    Wall.Height.NONE -> return null
                    Wall.Height.LOW -> wall.setHeight(sideFace, Wall.Height.TALL)
                    Wall.Height.TALL -> wall.setHeight(sideFace, Wall.Height.LOW)
                }
            }

            return if (wasUp) Sound.BLOCK_GRINDSTONE_USE else Sound.UI_STONECUTTER_TAKE_RESULT
        } else {
            // click side -> toggle side
            val sideFace = adjustedClickedFace ?: return null
            val hasFace = wall.getHeight(sideFace) != Wall.Height.NONE

            // Set height and choose sound
            if (hasFace) {
                wall.setHeight(sideFace, Wall.Height.NONE)
                return Sound.BLOCK_GRINDSTONE_USE
            } else {
                // Use opposite face's height, or low if there is nothing.
                var targetHeight = wall.getHeight(sideFace.oppositeFace)
                if (targetHeight == Wall.Height.NONE) {
                    targetHeight = Wall.Height.LOW
                }
                wall.setHeight(sideFace, targetHeight)
                return Sound.UI_STONECUTTER_TAKE_RESULT
            }
        }
    }

    /** Handles right-click block editing with the file tool. */
    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = true)
    fun onPlayerRightClick(event: PlayerInteractEvent) {
        if (!event.hasBlock() || event.action != Action.RIGHT_CLICK_BLOCK) {
            return
        }

        if (event.useItemInHand() == Event.Result.DENY) {
            return
        }

        // Require this custom file item.
        val player = event.player
        val hand = event.hand ?: return
        val item = player.equipment.getItem(hand)
        if (!isInstance(item)) {
            return
        }

        // Respect protections by simulating a block-break permission check.
        val block = event.clickedBlock ?: return
        val breakEvent = BlockBreakEvent(block, player)
        val server = module?.server ?: return
        server.pluginManager.callEvent(breakEvent)
        if (breakEvent.isCancelled) {
            return
        }

        val data = block.blockData
        val clickedFace = event.blockFace

        val sound: Sound?
        if (player.isSneaking) {
            if (data is Stairs) {
                sound = changeStairHalf(data)
            } else {
                return
            }
        } else {
            sound = when (data) {
                is MultipleFacing -> {
                    changeMultipleFacing(player, block, data, clickedFace)
                }

                is Wall -> {
                    changeWall(player, block, data, clickedFace)
                }

                is Stairs -> {
                    changeStairShape(player, block, data)
                }

                else -> {
                    return
                }
            }
        }

        // Abort if no state transition was applied.
        if (sound == null) {
            return
        }

        // Apply without physics updates to avoid cascading neighbor state changes.
        block.setBlockData(data, false)
        block.world.playSound(block.location, sound, SoundCategory.BLOCKS, 1.0f, 1.0f)

        // Consume durability and play swing animation.
        ItemUtil.damageItem(player, item, 1)
        PlayerUtil.swingArm(player, hand)
    }
}
